### ISSUE TITLE
`pane-focus-in` is happening in wrong (former) pane

### ISSUE THREAD

**Issue by ZlatyChlapec** [tmux/tmux#3506]
### Issue description

Expected behavior is to execute `echo` command in pane to which focus was switched. Instead of that it's sometimes getting triggered in pane from which focus was taken.

https://user-images.githubusercontent.com/6187861/227535040-23fc32ac-0086-48aa-8043-a15e05b2ff67.mp4

### Required information

Please provide the following information:

* tmux version (`tmux 3.3a`).
  * config:
```
# global session options
set-option -g default-command $SHELL # don't execute login shell
set-option -g history-limit 100000
set-option -g focus-events on

# global windows options
set-window-option -g mouse on

set-hook -g pane-focus-in 'run "echo "pane-focus-in"'
```
* Platform (`Darwin arm`) macOS Ventura 13.2.1 (22D68).
* $TERM inside and outside of tmux (`screen-256color`).
* alacritty version `alacritty 0.11.0 (8dbaa0b)`
  * config:
```
shell:
  program: /opt/homebrew/bin/fish
  args:
    - -c
    - 'tmux new \; set destroy-unattached'
```
* fish version `fish, version 3.6.0`

**Comment by Koston-0xDEADBEEF** [comment]
I've reproduced this bug on same OS and tmux version. As far as I can tell, it is not dependent on shell or terminal emulator, it happens also using two different terminal emulators as long as both have a tmux session using same tmux server.

**Comment by nicm** [comment]
Does this still happen now with tmux from master? There were some fixes around focus events not long ago.

**Comment by Koston-0xDEADBEEF** [comment]
The original `set-hook` no longer works:
`set-hook -g pane-focus-in 'run "echo "pane-focus-in"'`

The hook works when`echo` command output is redirected to a file, but stdout is not redirected to any pane.

FWIW, testing with following config, output is correct:
`set-hook -g pane-focus-in 'display-message "pane-focus-in #D"'`

**Comment by nicm** [comment]
OK so to be clear - do you mean that the original problem is fixed, and the `run-shell` problem is a new problem, or the original problem is *not* fixed?

**Comment by Koston-0xDEADBEEF** [comment]
> OK so to be clear - do you mean that the original problem is fixed, and the `run-shell` problem is a new problem

Yes - the `run-shell` problem is new, and original problem is very likely fixed.

**Comment by nicm** [comment]
`run-shell` will not longer write to a pane unless `-t` is give, so try this:
~~~~
set-hook -g pane-focus-in 'run -t: "echo "pane-focus-in"'
~~~~
I can't remember the reason for this change though...

**Comment by Koston-0xDEADBEEF** [comment]
Tested using the `-t` option, it *mostly* works. Output is redirected to either the correct pane, or in rare cases, nowhere.

I need to find some time to further narrow down the remaining problem where output isn't shown anywhere, should be done within next few days.

**Comment by Koston-0xDEADBEEF** [comment]
Sorry for the delay.

The case where output isn't directed to any pane can be reproduced with steps:

1) Trigger the hook in a pane, causing it to enter copy mode showing the text
2) Split the pane **without** exiting copy mode
3) Change focus to the new pane

After that, changing focus back to the old pane (still in copy mode) doesn't trigger anymore, until copy mode is exited.

**Comment by nicm** [comment]
I don't see this, it seems to work for me:

https://github.com/user-attachments/assets/50712c2b-6492-494e-9124-dd31b4d432ed
